backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala

```diff
@@ -0,0 +1,202 @@
+      goto(AwaitingJobResultAndAllHashes) using WithHashes(hashes)
+    case Event(CacheMissInternal(_), _) if !writeToCache =>
+      goto(UnupdatedCacheMiss)
+  }
+
+  when(AwaitingJobResultAndAllHashes) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache(returnCode, jobOutputs), WithHashes(hashes)) =>
+      writeEntryToDatabase(returnCode, jobOutputs)
+      sender ! CallCacheWriteInProgress
+      goto(WritingCacheEntry) using WithJobResult(returnCode, jobOutputs)
+  }
+
+  when(WritingCacheEntry) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(CheckWriteComplete, _) =>
+      sender ! CallCacheWriteInProgress
+      stay()
+    case Event(WriteCache(_, _), _) =>
+      // For the sake of idempotency, respond to this message too:
+      sender ! CallCacheWriteInProgress
+      stay()
+    case Event(CallCacheWriteSuccessInternal, oldData @ WithJobResult(_, _)) =>
+      goto(NewCacheEntryWritten) using oldData
+  }
+
+  when(FetchingJobResult) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheCheckInProgress
+      stay()
+    case Event(CacheHitResultInternal(returnCode, jobOutputs), WithHashes(hashes)) =>
+      goto(CacheLookupSuccess) using WithJobResult(returnCode, jobOutputs)
+  }
+
+  // Terminal States:
+
+  when(NewCacheEntryWritten) {
+    case Event(CheckCache, WithJobResult(returnCode, jobOutputs)) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache, _) =>
+      sender ! CallCacheWriteSuccess
+      stay()
+  }
+
+  when(UnreadCacheHit) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss // Let's just pretend we missed.
+      stay()
+    case Event(WriteCache, _) =>
+      sender ! CallCacheWriteSuccess // Pretend it was written. As far as we are concerned, there is an existing call cache entry for this call.
+      stay()
+  }
+
+  when(UnupdatedCacheMiss) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache, _) =>
+      sender ! CallCacheWriteSuccess // We probably shouldn't receive this but if we do we didn't fail... So the client should be happy.
+      stay()
+  }
+
+  when(CacheLookupSuccess) {
+    case Event(CheckCache, WithJobResult(returnCode, jobOutputs)) =>
+      sender ! CallCacheHit(SucceededResponse(backendJobDescriptor.key, returnCode, jobOutputs))
+      stay()
+    case Event(WriteCache, _) =>
+      sender ! CallCacheWriteSuccess // We didn't write it, but we didn't fail either. The client should be happy.
+      stay()
+  }
+
+  // If you never did anything, everything was a no-op.
+  when(NeverDidAnything) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheMiss
+      stay()
+    case Event(WriteCache(_, _), _) =>
+      sender ! CallCacheWriteSuccess
+      stay()
+  }
+
+  private def calculateHash(): Unit = {
+    // TODO: Do the work and then:
```

Comment:
If a backend chooses to use this call caching & except for runtime attributes and files, yes.